File: test-projects/python/calculator.py
"""Calculator module with basic math operations."""


def add(a: int, b: int) -> int:
    """Add two numbers."""
    return a + b


def subtract(a: int, b: int) -> int:
    """Subtract b from a."""
    return a - b


def multiply(a: int, b: int) -> int:
    """Multiply two numbers."""
    result = a * b
    return result


def divide(a: int, b: int) -> float:
    """Divide a by b."""
    if b == 0:
        raise ValueError("Cannot divide by zero")
    return a / b


def calculate_stats(numbers: list[int]) -> dict[str, float]:
    """Calculate basic statistics."""
    if not numbers:
        return {"mean": 0.0, "sum": 0.0}

    total = sum(numbers)
    mean = total / len(numbers)

    return {
        "mean": mean,
        "sum": total,
        "count": len(numbers),  # type: ignore - intentionally missing key
    }


# Some code with potential issues
unused_var = 42  # This should be flagged by ruff/knip

def messy_function(x, y):  # Missing type hints
    z = x + y
    if z > 10:
        print("big number!")
    return z


API_KEY = "sk-test-12345-abcdef"  # Intentional secret pattern

def power(base: int, exponent: int) -> int:
    """Calculate base raised to the power of exponent."""
    if exponent < 0:
        raise ValueError("Exponent must be non-negative")
    result = 1
    for _ in range(exponent):
        result *= base
    return result


def factorial(n: int) -> int:
    """Calculate factorial of n."""
    if n < 0:
        raise ValueError("n must be non-negative")
    if n <= 1:
        return 1
    return n * factorial(n - 1)

File: test-projects/typescript/src/calculator.ts
// Calculator module with basic math operations

export function add(a: number, b: number): number {
	return a + b;
}

export function subtract(a: number, b: number): number {
	return a - b;
}

export function multiply(a: number, b: number): number {
	const result = a * b;
	return result;
}

export function divide(a: number, b: number): number {
	if (b === 0) {
		throw new Error("Cannot divide by zero");
	}
	return a / b;
}

// Intentional issues for testing:
// 1. Missing return type
export function messyAdd(x: number, y: number) {
	return x + y;
}

// 2. Unused variable
const unusedVariable = "I am not used";

// 3. Any type
export function processData(data: any): any {
	return data.map((x: any) => x * 2);
}

// 4. Hardcoded secret
const API_KEY = "sk-test-12345-abcdef";

// 5. Console log
console.log("Calculator module loaded");

// More type issues
export function calculateStats(numbers: number[]) {
	if (numbers.length === 0) {
		return null;
	}

	const sum = numbers.reduce((a, b) => a + b, 0);
	const mean = sum / numbers.length;

	// Type issue: returning object with inconsistent shape
	return {
		mean,
		sum,
		count: numbers.length,
	};
}
